File: c++/leetcode/20. Valid Parentheses.cpp
class Solution
{
public:
  bool isValid(string s)
  {
    if (s.length() % 2 != 0)
      return false;

    unordered_map<char, char> parenthesis = {
        {'(', ')'},
        {'{', '}'},
        {'[', ']'}};

    stack<char> seen;

    for (auto i : s)
    {
      if (parenthesis.find(i) != parenthesis.end())
        seen.push(i);
      else
      {
        if (seen.empty() || i != parenthesis[seen.top()])
          return false;
        seen.pop();
      }
    }

    return seen.empty();
  }
};

File: javascript/leetcode/20. Valid Parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  if (s.length % 2 !== 0) return false;

  const parenthesis = new Map([
    ["(", ")"],
    ["{", "}"],
    ["[", "]"],
  ]);

  const seen = [];

  for (let i of s) {
    if (parenthesis.has(i)) seen.push(i);
    else {
      if (seen.length === 0 || i != parenthesis.get(seen[seen.length - 1]))
        return false;
      seen.pop();
    }
  }

  return seen.length === 0;
};
